coordinated tombstone removal end2end test

bashtanov opened the pull request:
Make sure tombstone removal is not subject to race with recovery anymore

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
rptest/services/redpanda: utility context manager for pausing a node

bashtanov committed:
rptest/tombstone_removal_e2e_test: coordinated tombstone removal test

Make sure tombstone removal is not subject to race with recovery anymore

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75240-->
<!--c9eec3b8-f15e-4d78-8f5e-431b185fae50-->
<!--title-->
## Retry command for [Build#75240](https://buildkite.com/redpanda/redpanda/builds/75240)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/tombstone_removal_e2e_test.py::TombstoneRemovalTest.test_single_slow_follower
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--58842337-2645-4196-9551-cf95b4c301ca-->
<!--TEST RESULTS-->
<!--23045a0a-cfb0-423c-8908-ea056094954d-->
<!--TEST RESULTS-->
<!--b9f54d1e-dfb6-4bfc-b4c7-e6b41ec9aeac-->
<!--TEST RESULTS-->
<!--8d0d2ff2-b999-403a-b670-cf4283e6c86d-->
<!--TEST RESULTS-->
<!--592da435-91c5-4ffd-a239-92c4967e84f7-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75240">build#75240</a></summary>


| test_class                          | test_method                                | test_arguments                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:------------------------------------|:-------------------------------------------|:-----------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest               | test_creating_and_listing_migrations       | null                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75240#019a312c-9dea-4c42-9058-ebfc6d9bd384 | FLAKY         | 17/21    | upstream reliability is '98.85931558935361'. current run reliability is '80.95238095238095'. drift is 17.90693 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations   |
| DeleteRecordsTest                   | test_delete_records_concurrent_truncations | {"cloud_storage_enabled": true, "truncate_point": "at_high_watermark"} | integration | https://buildkite.com/redpanda/redpanda/builds/75240#019a3133-5680-4c6a-86a1-1be4aa49431c | FLAKY         | 17/21    | upstream reliability is '97.11751662971176'. current run reliability is '80.95238095238095'. drift is 16.16514 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_concurrent_truncations |
| TombstoneRemovalTest                | test_single_slow_follower                  | null                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75240#019a3133-5679-47f2-b0cb-52f7b9b1f44a | FLAKY         | 19/21    | The test was found to be new, and no failures are allowed                                                                                                               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TombstoneRemovalTest&test_method=test_single_slow_follower               |
| WriteCachingFailureInjectionE2ETest | test_crash_all                             | {"use_transactions": false}                                            | integration | https://buildkite.com/redpanda/redpanda/builds/75240#019a312c-9de8-4b04-a7c2-e84ee517e645 | FLAKY         | 16/21    | upstream reliability is '83.96421197522368'. current run reliability is '76.19047619047619'. drift is 7.77374 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75482">build#75482</a></summary>


| test_class                           | test_method                       | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                  |
|:-------------------------------------|:----------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest            | test_write                        | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75482#019a4a47-f470-4a43-bf7a-315ccd2fe8c1 | FLAKY         | 14/21    | upstream reliability is '91.91402251791197'. current run reliability is '66.66666666666666'. drift is 25.24736 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                            |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync        | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "with_failures": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/75482#019a4a50-1d9c-4093-ba12-4895c474fded | FLAKY         | 17/21    | upstream reliability is '91.14331723027375'. current run reliability is '80.95238095238095'. drift is 10.19094 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync        | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": false}                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/75482#019a4a50-1d9e-4ae7-97bb-9526217f5235 | FLAKY         | 15/21    | upstream reliability is '94.96268656716418'. current run reliability is '71.42857142857143'. drift is 23.53412 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| ShadowLinkingReplicationTests        | test_replication_timestamps_match | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "timestamp_type": "CreateTime"} | integration | https://buildkite.com/redpanda/redpanda/builds/75482#019a4a47-f470-4a43-bf7a-315ccd2fe8c1 | FLAKY         | 20/21    | upstream reliability is '95.56650246305419'. current run reliability is '95.23809523809523'. drift is 0.32841 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| MountUnmountIcebergTest              | test_simple_remount               | {"cloud_storage_type": 1}                                                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/75482#019a4a47-f471-470e-982a-1e4fc3be87ec | FLAKY         | 17/21    | upstream reliability is '76.40653357531761'. current run reliability is '80.95238095238095'. drift is -4.54585 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                     |
| WriteCachingFailureInjectionE2ETest  | test_crash_all                    | {"use_transactions": false}                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/75482#019a4a47-f470-4a43-bf7a-315ccd2fe8c1 | FLAKY         | 19/21    | upstream reliability is '86.63558663558663'. current run reliability is '90.47619047619048'. drift is -3.8406 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all              |
| TxUpgradeCompactionTest              | upgrade_with_compaction_test      | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75482#019a4a47-f46f-4301-aff0-68f55a94662d | FLAKY         | 19/21    | upstream reliability is '98.93617021276596'. current run reliability is '90.47619047619048'. drift is 8.45998 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxUpgradeCompactionTest&test_method=upgrade_with_compaction_test            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75533">build#75533</a></summary>


| test_class                          | test_method                       | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                  |
|:------------------------------------|:----------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests       | test_replication_timestamps_match | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "timestamp_type": "CreateTime"} | integration | https://buildkite.com/redpanda/redpanda/builds/75533#019a4e6f-2a7c-41d7-86d8-e5d5588598cd | FLAKY         | 20/21    | upstream reliability is '95.6896551724138'. current run reliability is '95.23809523809523'. drift is 0.45156 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| MountUnmountIcebergTest             | test_simple_remount               | {"cloud_storage_type": 1}                                                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/75533#019a4e6f-2a7d-434f-a822-fb4d041aeea7 | FLAKY         | 15/21    | upstream reliability is '76.12208258527828'. current run reliability is '71.42857142857143'. drift is 4.69351 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                     |
| PartitionMovementTest               | test_static                       | {"num_to_upgrade": 0}                                                                                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/75533#019a4e76-08ff-4e98-8b9e-b449374f8740 | FLAKY         | 19/21    | upstream reliability is '86.1788617886179'. current run reliability is '90.47619047619048'. drift is -4.29733 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                               |
| ShadowLinkingRandomOpsTest          | test_node_operations              | {"failures": true}                                                                                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/75533#019a4e6f-2a7d-434f-a822-fb4d041aeea7 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                 |
| WriteCachingFailureInjectionE2ETest | test_crash_all                    | {"use_transactions": false}                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/75533#019a4e6f-2a7c-41d7-86d8-e5d5588598cd | FLAKY         | 18/21    | upstream reliability is '87.59430008382229'. current run reliability is '85.71428571428571'. drift is 1.88001 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75652">build#75652</a></summary>


| test_class                          | test_method    | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                     |
|:------------------------------------|:---------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| WriteCachingFailureInjectionE2ETest | test_crash_all | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75652#019a5489-8eb5-46da-a7cc-a3b01540cd5b | FLAKY         | 20/21    | upstream reliability is '89.12058023572075'. current run reliability is '95.23809523809523'. drift is -6.11752 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75876">build#75876</a></summary>


| test_class                    | test_method                                  | test_arguments                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:------------------------------|:---------------------------------------------|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_topic_delete                            | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75876#019a6121-3ad3-4d26-8951-a90b27de6619 | FLAKY         | 20/21    | upstream reliability is '98.13084112149532'. current run reliability is '95.23809523809523'. drift is 2.89275 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete            |
| SegmentMsTest                 | test_segment_rolling_with_retention_consumer | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75876#019a6121-3ace-445c-b315-ee1833312de9 | FLAKY         | 20/21    | upstream reliability is '93.70229007633588'. current run reliability is '95.23809523809523'. drift is -1.53581 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
</details>
<!--/body-->

dotnwat commented:
/ci-repeat 5
dt-repeat 5
skip-units
skip-redpanda-build
tests/rptest/tests/tombstone_removal_e2e_test.py

joe-redpanda reviewed (dismissed):
rad

bashtanov committed:
rptest/services/kgo_verfier/producer: typify .oneshot()

bashtanov committed:
rptest/services/admin: typify a few functions

list_reconfigurations,
get_partition,
set_partition_replicas,
transfer_leadership_to

joe-redpanda commented on tests/rptest/services/admin.py:
```diff
@@ -1092,7 +1092,7 @@ def trigger_cores_rebalance(self, node):
         return self._request("post", path, node=node)
 
-    def list_reconfigurations(self, node=None):
+    def list_reconfigurations(self, node: ClusterNode | None = None):
```
nit: Optional[ClusterNode]

bashtanov replied:
I'm not sure it's any better. `|` syntax was introduced after `Optional`, and PEP 604 that introduces it suggests that `|` can now be used instead of `Optional`. There's no recommendation or deprecation though. We have hundreds of occurrences of each syntax in the codebase already, so if we want to unify it I'd discuss it on slack first.

joe-redpanda commented on tests/rptest/tests/tiered_storage_enable_test.py:
```diff
@@ -55,7 +55,7 @@ def start_producer(self):
         self.producer.start(clean=False)
         self.producer.wait_for_acks(
-            5 * (self.producer_throughput / self.msg_size), 120, 1
+            5 * self.producer_throughput // self.msg_size + 1, 120, 1
```
q:
one got floor(a/b)*5
the other got floor(a/b)*5+1

any reason for the difference?

bashtanov replied:
you're right, I'll remove the `+ 1`

joe-redpanda reviewed (dismissed):
nit, question, and copilot comments.
otherwise lgtm

bashtanov commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+        data59 = {
+            line
+            for line in lines
+            if "key-5=" <= line[0:6] <= "key-9="
```
I think it's more readable the way it is. All key numbers are single-digit anyway.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75533-->
<!--82a7390e-b04c-4ac3-b2b2-5a704a949825-->
<!--title-->
## Retry command for [Build#75533](https://buildkite.com/redpanda/redpanda/builds/75533)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

nvartolomei commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+            key_set_cardinality=key_set_cardinality,
+            tombstone_probability=0.0,
+        )
+        try:
```
What's the intention here? Do you want to send all messages or 100?

If the former why you set high msg_count and not just 100? You can also just `KgoVerifierProducer.oneshot([params from above as-is])`

If the later, you need to call `wait()` not `stop()` I believe.

bashtanov replied:
I want to send all messages, and wait for acks for no longer than 100 seconds. I'll see if I can simplify it using oneshot, thanks

nvartolomei replied:
Ah. Weird to see wait_for_acks when you want to wait for all messages (i assumed 100 means number of acks by mistake). oneshot or wait() is more readable.

wait_for_acks is often used to say send unbounded messages but you want only for a small number to be delivered before continuing with the test to make sure that data is flowing.

nvartolomei commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+        return False
+
+    def transfer_leadership_to_node(self, node_id: int):
+        wait_until(
```
Why not paste the body of `leadership_transferred_to_node` with wait_until at the end? Why are we calling transfer api in a loop? Leave a note why the code was written in this non-intuitive way.

nvartolomei commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+        )
+
+    def get_data59_from_node(self, node_id: int):
+        self.transfer_leadership_to_node(node_id)
```
although the balancer is disabled, the leadership may still move due to network latency, right? i.e. in cdt. For a bulletproof test it needs to assert at the end that leadership/epoch didn't change in-between.

You are expecting leadership to be stable but this expectation can be broken in too many ways in the future and by new settings... (like [this](https://github.com/redpanda-data/redpanda/pull/28331) one)

nvartolomei commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+            err_msg=f"timed out waiting for leadership transfer to node {node_id}",
+        )
+
+    def get_data59_from_node(self, node_id: int):
```
nit: small code doc?

nvartolomei commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+        return sum(1 for o in offsets if o < offset) <= n
+
+    @cluster(num_nodes=6)  # 5 for cluster + 1 for producer
+    def test_single_slow_follower(self):
```
please add some prose code doc to describe the scenario

helpful for readers, helpful for test results.html to quickly know what the test is doing and triage failures for someone less familiar with the test

nvartolomei commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+        )
+
+        # let node 4 recover at least some data but no tombstones
+        def data_recovered_halfway() -> bool:
```
this code seems racy and easy to make the test do the wrong thing... i don't have good suggestions for how to make it more robust but I'd like to ask at least to re-run this check after the node is paused

actually, can we maybe look at follower stats or something for recovery progress? maybe network partition it so that we can still query admin apis for dirty offset instead of pausing?

bashtanov replied:
Are you worried node4 recovers 64Mb of data in 1 second? We set `raft_learner_recovery_rate` to 10Mb/s to prevent this.

nvartolomei replied:
There is nothing guaranteeing that only 1 second passed between check and node getting paused. Also I don't trust throttling and who knows at what granularity it is implemented/will be in the future.

bashtanov replied:
The check is run every second. Once it passes the node is signalled to pause.

We routinely rely on recovery throttling in tests.

joe-redpanda replied:
If this test becomes flakey in the future i feel like this would be a good adminv2/internal api to jam a replica at a certain offset

nvartolomei commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+
+            # make sure node 1 is not compacting away tombstones until node 4
+            # catches up and compacts away correspondent data records first
+            time.sleep(30)
```
For this assertion to be robust this test needs to be run twice. Once with node4 paused where we assert that compaction of tombstones didn't happen and once with node4 active where we assert that compaction of tombstones does happen. This way you get high confidence that sleeping for 30 seconds actually is enough for compaction process to be triggered. Without this, there is no evidence that any compaction will actually be scheduled. Maybe you tested it manually and it worked when you did but there are no guarantees this will work tomorrow.

That's the best way I know for asserting "safety properties" i.e. that nothing bad happens. You need proof that the bad thing actually had a chance to happen.

bashtanov replied:
> Without this, there is no evidence that any compaction will actually be scheduled.

What is the scenario you'd like to additionally test? Same as here but no leadership transfers?

nvartolomei replied:
Once with node4 paused where we assert that compaction of tombstones didn't happen **and once with node4 active where we assert that compaction of tombstones does happen. This way you get high confidence that sleeping for 30 seconds actually is enough for compaction process to be triggered.**

> compaction process to be triggered

and tombstones deleted

nvartolomei commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+            # make sure node 1 is not compacting away tombstones until node 4
+            # catches up and compacts away correspondent data records first
+            time.sleep(30)
+            abnormal_compaction = self.no_more_than_n_records_below_offset(
```
Don't understand this. We assert that there are 0 records below `offset_after_tombstones`? Where are the tombstones? 🤔

nvartolomei replied:
And why in the next check we check that there are less than 10 offests?

bashtanov replied:
> We assert that there are 0 records below offset_after_tombstones?

No, we record the result of this check into a boolean variable and later on assert its value is `False`

> And why in the next check we check that there are less than 10 offests?

For to make sure at least the data records before the tombstones have  been removed

nvartolomei replied:
Ok. I miss-read the code 🤦

```
abnormal_compaction = self.no_more_than_n_records_below_offset(
0, offset_after_tombstones)
```

May I suggest renaming `abnormal_compaction` to `tombstones_gone`?

bashtanov replied:
I'll rename it to `tombstones_incorrectly_removed` to keep stressing it's not expected

nvartolomei commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+            abnormal_compaction = self.no_more_than_n_records_below_offset(
+                0, offset_after_tombstones
+            )
+            if abnormal_compaction:
```
why we don't assert right here?

bashtanov replied:
For this test to demonstrate the data corruption if run against an older version of redpanda. I'll add a comment

nvartolomei reviewed (commented):
a bit confused. see first the comments toward end of the test please

nvartolomei commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,283 @@
+        # compact node 1
+        wait_until(
+            lambda: self.no_more_than_n_records_below_offset(
+                10, offset_after_tombstones
```
Should this assert for "no more than 5"? You wrote additional keys for 0-4 after tombstones so they will get compacted I guess?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75652-->
<!--61e9dea0-6ad2-43a2-be71-5ea1356ba0d4-->
<!--title-->
## Retry command for [Build#75652](https://buildkite.com/redpanda/redpanda/builds/75652)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_smoke_test.py::RedpandaNodeOperationsSmokeTest.test_node_ops_smoke_test@{"cloud_storage_type":1,"mixed_versions":true}
```
<!--/body-->

bashtanov commented:
the failure is not related and has been recently fixed

joe-redpanda commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,267 @@
+
+    def has_59_data_on_node(self, node: ClusterNode) -> bool:
+        """Returns non-tombstone data records for keys 5-9 from given node"""
+        viewer = OfflineLogViewer(self.redpanda)
```
I didn't realize we had this.

bashtanov replied:
I only learnt about it recently

joe-redpanda commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,267 @@
+        tombstone_offsets = [
+            self.rpk.produce(
+                topic=self.topic_name,
+                key="key-{}".format(key),
```
nit:
quietly depends on the keys emitted by kgo, worth a comment

joe-redpanda commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,267 @@
+            assert bytes_moved <= total_bytes_to_move * 0.75, "node recovered too fast"
+            return bytes_moved >= total_bytes_to_move * 0.25
+
+        wait_until(
```
nit:
worth a bit of math to set the timeout expectation
256 * 1024 * 1024 / (10*1024*1024)  -> 26ish seconds in ideal circumstances

bashtanov replied:
I'll put these as a comment but will keep timeout as is to allow for CI nodes slowness

joe-redpanda commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,267 @@
+
+            # make sure node 1 is not compacting away tombstones until node 4
+            # catches up and compacts away correspondent data records first
+            time.sleep(30)
```
could we just do a wait until get_first_offset hits the tombstone offsets

bashtanov replied:
We'll then need to give it a bit more time to make sure it doesn't incorrectly remove these tombstones

joe-redpanda commented on tests/rptest/tests/tombstone_removal_e2e_test.py:
```diff
@@ -0,0 +1,267 @@
+        )
+
+        # make node 4 the leader, thus making sure it caught up
+        self.transfer_leadership_to_node(4)
```
We're transferring leadership which at some point will have to wait for 4 to catch up all the way. Raise LRR back to max?

joe-redpanda reviewed (dismissed):
nitpicks and commentary
the core logic looks good to me though

bashtanov commented:
last push was a rebase till new upstream/dev only

nvartolomei reviewed (commented):
Everything looks good, except this bit which is still unaddressed (previous thread https://github.com/redpanda-data/redpanda/pull/28261#discussion_r2491007061):

```py
# but make sure node 1 is not compacting away tombstones until node 4
# catches up and compacts away correspondent data records first
time.sleep(10)
```

We need to find a better condition to wait for here. Either as I suggested before with a test (matrix or standalone) that proves that in 10 seconds regular compaction (without lagging replica) does need remove tombstones for the same cluster/topic configuration.

Or, maybe counting log lines before and after with a text like (invented/didn't check implementation) "compaction refusing tombstone deletion".

Not sure how I can trust the test otherwise.

bashtanov commented:
@nvartolomei the waiting time of 10s covers the following:
1) leader request followers to report their MCCOs once every 1/24 of `tombstone_retention_ms` (~42ms in this test),
2) upon receiving an MCCO, leader waits for 3s (hardcoded) before recalculating MTRO and distributing it to followers again,
3) we allow for network glitches and node slowness.

Waiting for specific messages in the log would assume a certain scenario, while in reality they may be different: RPC replies from nodes come in different order, later RPC iterations cancel the earlier ones etc. The test checks that even after multiple rounds of attempted coordination illegal tombstone removal is not allowed.

Would it convince you if I add the calculation above as a code comment?

nvartolomei commented:
@bashtanov you are making lots of assumptions including compaction scheduling. Also these assumptions are made very far from the implementation which increases the chance of things drifting.

We ended up multiple times with tests which made similar assumptions where they ended up being completely useless after the implementation changed and tests continued to pass without actually verifying anything.

bashtanov commented:
I understand your concerns, but there's no explicit event we can waiting for instead of a sleep, not even a log message. Coordinated compaction does not operate strict rounds, as it should be resilient w.r.t. node outages. The leader receives new MCCOs from followers, but at no point it is aware that it has refreshed all of them since some point in the past. Today we only can check that
1) for each follower (apart from node 4), the follower's MCCO has been recalculated, and either didn't change or was sent to leader;
2) and, if it was sent to leader the leader has received it and recalculated MTRO based on this.

That'd be waiting on multiple messages from multiple nodes and linking them together. It'd essentially duplicate the core logic we're testing. I think it would be a stronger dependency on implementation details than relying on timings.
